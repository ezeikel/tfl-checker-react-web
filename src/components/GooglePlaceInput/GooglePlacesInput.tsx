import PlacesAutocomplete, {
  geocodeByAddress,
  getLatLng,
} from "react-places-autocomplete";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import {
  Input,
  Loading,
  SuggestionsWrapper,
  Suggestion,
  PlacesByGoogle,
  List,
} from "./GooglePlacesInput.styled";
import { useJourneyContext } from "../../contexts/journey";

type GooglePlaceInputProps = {
  inputId: string;
  placeholder: string;
  address?: string;
};

const GooglePlacesInput = ({
  inputId,
  placeholder,
  address = "",
}: GooglePlaceInputProps) => {
  const { setFrom, setTo } = useJourneyContext();

  const handleSelect = async (add: string) => {
    const [results] = await geocodeByAddress(add);
    const latLng = await getLatLng(results);
    if (inputId === "from") {
      setFrom((prev) => ({
        ...prev,
        address: add,
        coordinates: {
          latitude: latLng.lat,
          longitude: latLng.lng,
        },
      }));
    } else {
      setTo((prev) => ({
        ...prev,
        address: add,
        coordinates: {
          latitude: latLng.lat,
          longitude: latLng.lng,
        },
      }));
    }
  };

  const handeChange = (add: string) => {
    if (inputId === "from") {
      setFrom((prev) => ({
        ...prev,
        address: add,
      }));
    } else {
      setTo((prev) => ({
        ...prev,
        address: add,
      }));
    }
  };

  const searchOptions = {
    componentRestrictions: { country: ["gb"] },
  };

  return (
    <PlacesAutocomplete
      value={address}
      onChange={handeChange}
      onSelect={handleSelect}
      searchOptions={searchOptions}
    >
      {({ getInputProps, suggestions, getSuggestionItemProps, loading }) => (
        <>
          <Input
            // eslint-disable-next-line react/jsx-props-no-spreading
            {...getInputProps({
              inputId,
              placeholder,
            })}
          />
          <SuggestionsWrapper>
            {loading ? (
              <Loading>
                <FontAwesomeIcon
                  icon={["fad", "spinner-third"]}
                  color="var(--color-background)"
                  size="2x"
                />
              </Loading>
            ) : null}
            <List>
              {suggestions.map((suggestion) => {
                return (
                  <Suggestion
                    active={suggestion.active}
                    // eslint-disable-next-line react/jsx-props-no-spreading
                    {...getSuggestionItemProps(suggestion)}
                  >
                    {suggestion.description}
                  </Suggestion>
                );
              })}
              <PlacesByGoogle>
                Places by <strong>Google</strong>
              </PlacesByGoogle>
            </List>
          </SuggestionsWrapper>
        </>
      )}
    </PlacesAutocomplete>
  );
};

export default GooglePlacesInput;
