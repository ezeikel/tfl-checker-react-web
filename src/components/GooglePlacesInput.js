import React from "react";
import PropTypes from "prop-types";
import PlacesAutocomplete, {
  geocodeByAddress,
  getLatLng,
} from "react-places-autocomplete";
import styled from "styled-components";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { rotate } from "../GlobalStyle";

const Input = styled.input`
  font-size: 1.6rem;
  width: 100%;
  outline: none;
  border: 0;
  background-color: transparent;
  font-weight: bold;
  ::-webkit-input-placeholder {
    /* Edge */
    color: #aeb8c3;
    font-weight: bold;
  }

  :-ms-input-placeholder {
    /* Internet Explorer 10-11 */
    color: #aeb8c3;
    font-weight: bold;
  }

  ::placeholder {
    color: #aeb8c3;
    font-weight: bold;
  }
`;

const Loading = styled.div`
  display: flex;
  padding: 8px 0;
  animation: ${rotate} 1.2s linear infinite;
  align-self: flex-start;
`;

const SuggestionsWrapper = styled.div`
  display: flex;
  flex-direction: column;
  [aria-expanded="true"] + & {
    margin-top: 8px;
  }
`;

const Suggestion = styled.div`
  font-family: var(--primary-font-family);
  font-size: 1.6rem;
  background-color: ${props =>
    `var(--color-${props.active ? "background" : "white"})`};
  color: ${props => `var(--color-${props.active ? "white" : "black"})`};
  cursor: pointer;
  padding: 8px;
`;

const PlacesByGoogle = styled.div`
  font-size: 11px;
  background-color: white;
  padding: 8px;
  display: none;
  &:not(:only-child) {
    display: block;
  }
`;

const List = styled.div`
  > div {
    & + div {
      border-top: 1px solid #eeeeee;
    }
  }
`;

const GooglePlacesInput = ({
  inputId,
  placeholder,
  address,
  setAddress,
  setLocation,
}) => {
  const handleSelect = async value => {
    const [results] = await geocodeByAddress(value);
    const latLng = await getLatLng(results);

    setAddress(value);
    setLocation(latLng);
  };

  const searchOptions = {
    componentRestrictions: { country: ["gb"] },
  };

  return (
    <PlacesAutocomplete
      value={address}
      onChange={setAddress}
      onSelect={handleSelect}
      searchOptions={searchOptions}
    >
      {({ getInputProps, suggestions, getSuggestionItemProps, loading }) => (
        <>
          <Input
            // eslint-disable-next-line react/jsx-props-no-spreading
            {...getInputProps({
              inputId,
              placeholder,
            })}
          />
          <SuggestionsWrapper>
            {loading ? (
              <Loading>
                <FontAwesomeIcon
                  icon={["fad", "spinner-third"]}
                  color="var(--color-background)"
                  size="2x"
                />
              </Loading>
            ) : null}
            <List>
              {suggestions.map(suggestion => {
                return (
                  <Suggestion
                    active={suggestion.active}
                    // eslint-disable-next-line react/jsx-props-no-spreading
                    {...getSuggestionItemProps(suggestion)}
                  >
                    {suggestion.description}
                  </Suggestion>
                );
              })}
              <PlacesByGoogle>
                Places by <strong>Google</strong>
              </PlacesByGoogle>
            </List>
          </SuggestionsWrapper>
        </>
      )}
    </PlacesAutocomplete>
  );
};

GooglePlacesInput.defaultProps = {
  address: "",
};

GooglePlacesInput.propTypes = {
  inputId: PropTypes.string.isRequired,
  placeholder: PropTypes.string.isRequired,
  address: PropTypes.string,
  setAddress: PropTypes.func.isRequired,
  setLocation: PropTypes.func.isRequired,
};

export default GooglePlacesInput;
