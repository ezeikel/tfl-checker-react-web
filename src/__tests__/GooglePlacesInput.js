import React from 'react';
import { render } from '@testing-library/react';
import GooglePlacesInput from '../components/GooglePlacesInput';

const setupGoogleMock = () => {
  // Mock Google Maps JavaScript API
  const google = {
    maps: {
      places: {
        AutocompleteService: class{},
        PlacesServiceStatus: {
          INVALID_REQUEST: 'INVALID_REQUEST',
          NOT_FOUND: 'NOT_FOUND',
          OK: 'OK',
          OVER_QUERY_LIMIT: 'OVER_QUERY_LIMIT',
          REQUEST_DENIED: 'REQUEST_DENIED',
          UNKNOWN_ERROR: 'UNKNOWN_ERROR',
          ZERO_RESULTS: 'ZERO_RESULTS',
        },
      },
      Geocoder: () => {},
      GeocoderStatus: {
        ERROR: 'ERROR',
        INVALID_REQUEST: 'INVALID_REQUEST',
        OK: 'OK',
        OVER_QUERY_LIMIT: 'OVER_QUERY_LIMIT',
        REQUEST_DENIED: 'REQUEST_DENIED',
        UNKNOWN_ERROR: 'UNKNOWN_ERROR',
        ZERO_RESULTS: 'ZERO_RESULTS',
      },
    },
  };
  global.window.google = google;
};

beforeAll(() => {
  setupGoogleMock();
});

let mockProps;

beforeEach(() => {
  mockProps = {
    inputId: "id",
    placeholder: "Placeholder",
    setAddress: jest.fn(),
    setLocation: jest.fn()
  };
});

test('renders without crashing', () => {
  const { container } = render(
    <GooglePlacesInput {...mockProps} /> // eslint-disable-line react/jsx-props-no-spreading
  );
  expect(container).toMatchSnapshot();
});