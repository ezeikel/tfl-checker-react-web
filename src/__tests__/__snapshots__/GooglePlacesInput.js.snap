// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<div>
  <input
    aria-autocomplete="list"
    aria-expanded="false"
    autocomplete="off"
    class="sc-AxjAm hhXapv"
    placeholder="Placeholder"
    role="combobox"
    type="text"
    value=""
  />
  <div
    class="sc-AxiKw jAmDvQ"
  >
    <div
      class="sc-AxgMl jGFEmy"
    >
      <div
        class="sc-AxhUy ixiydF"
      >
        Places by 
        <strong>
          Google
        </strong>
      </div>
    </div>
  </div>
</div>
`;
